src/main/java/com/aws/greengrass/tes/HttpServerImpl.java

```diff
@@ -5,51 +5,203 @@
 
     /**
      * Constructor.
-     * @param port Http server port
+     * @param port HTTP server port
      * @param credentialRequestHandler request handler for server requests
      * @param executorService executor service instance
-     * @throws IOException When server creation fails
      */
-    HttpServerImpl(int port, HttpHandler credentialRequestHandler, ExecutorService executorService) throws IOException {
-        httpImpl = HttpServer.create(new InetSocketAddress("localhost", port), 0);
+    HttpServerImpl(int port, HttpHandler credentialRequestHandler, ExecutorService executorService) {
+        this(port, credentialRequestHandler, executorService,
+                HttpServerProvider.provider(), SocketFactory.getDefault());
+    }
+
+    HttpServerImpl(int port, HttpHandler credentialRequestHandler, ExecutorService executorService,
+                   HttpServerProvider httpServerProvider, SocketFactory socketFactory) {
         this.credentialRequestHandler = credentialRequestHandler;
         this.executorService = executorService;
+        this.configuredPort = port;
+        this.httpServerProvider = httpServerProvider;
+        this.socketFactory = socketFactory;
     }
 
     @Override
-    public void start() {
-        httpImpl.createContext(URL, credentialRequestHandler);
-        httpImpl.setExecutor(executorService);
-        httpImpl.start();
+    public void start() throws IOException, InterruptedException {
+        createServers(configuredPort);
+        getServers().forEach(this::start);
+    }
+
+    private void start(HttpServer server) {
+        server.createContext(URL, credentialRequestHandler);
+        server.setExecutor(executorService);
+        server.start();
     }
 
     @Override
     public void stop() {
-        httpImpl.stop(TIME_TO_WAIT_BEFORE_SHUTDOWN_IN_SECONDS);
+        getServers().forEach(this::stop);
+    }
+
+    private void stop(HttpServer server) {
+        server.stop(TIME_TO_WAIT_BEFORE_SHUTDOWN_IN_SECONDS);
+    }
+
+    List<InetSocketAddress> getServerAddresses() {
+        return getServers().stream()
+                .map(HttpServer::getAddress)
+                .collect(Collectors.toList());
     }
 
     int getServerPort() {
-        return httpImpl.getAddress().getPort();
+        return getServers().stream()
+                .findFirst()
+                .map(HttpServer::getAddress)
+                .map(InetSocketAddress::getPort)
+                .orElse(PORT_UNKNOWN);
+    }
+
+    private List<HttpServer> getServers() {
+        return Stream.of(serverIPv4, serverIPv6)
+                .filter(Objects::nonNull)
+                .collect(Collectors.toList());
+    }
+
+    private void createServers(int port) throws IOException, InterruptedException {
+        if (!getServers().isEmpty()) {
+            logger.atInfo().kv("port", port).log("Skipping server creation, servers already running");
+            return;
+        }
```

Comment:
How does this work when we need to restart if some configuration for TES has changed?

Reply:
new `HttpServerImpl` is created on each `startup`, https://github.com/aws-greengrass/aws-greengrass-nucleus/pull/1351/files#diff-53c817ca085df370edb8025c1ae9981c0e0d62c164ade69a91290289b3e7d573R95, so it'll respond to config changes.  Technically this path won't ever get triggered today, more of a protection against if `HttpServerImpl#start` were called twice with the same instance.

Reply:
I see, but looks like that happens on service startup for TES, is config expected to change and take effect without service restart? (I know this is not part of this change and how it existed, just curious how this currently works and if have scope for improvement)

Reply:
Ahh, yeah I was just focused on keeping the behavior similar to before, port change taking effect after restart.  I don't see why we couldn't react to port changes in the future.  In the original PR for TES there's a [TODO](https://github.com/aws-greengrass/aws-greengrass-nucleus/pull/240/files#r424732402) for it. Although there's a comment in the code rn saying that [port change shouldn't be allowed](https://github.com/aws-greengrass/aws-greengrass-nucleus/commit/fa3cf396d0001602e291070c970f568f4534eec4#diff-c2b86c6fe185344327806e5cb614804c7e81c6abd7ee63b41f42384fb28d4238R58), curious what that's about